Intercept links to youtube videos and playlists to open them in the app.

atomjack opened the pull request:
Currently, when clicking on a link (in the video description) to a YouTube Video, or Playlist, it attempts to open the link in the device's browser. This PR instead opens the link in the app itself (if it's a link to a video or playlist). It also captures long clicks on any link, and shows a menu allowing the user to Open the link in the devices browser, copy url, or share.

I chose to put the method that alters the TextView (via the Link Builder library I added to build.gradle) in SkyTubeApp, so that it could be used from anywhere, although it's currently only used from within `YouTubePlayerFragment`.

Also, there's an unused YouTubePlaylistListener interface, but I figured I'd leave it in in case it was ever needed (I could remove it though).

Here are two videos you can test it on:
https://www.youtube.com/watch?v=NmM9HA2MQGI - Has a few links to videos
https://www.youtube.com/watch?v=_GzE99AmAQU - Has a link to a playlist

ram-on commented:
Hello @atomjack, I've tested the PR and I really like the idea.  The following is my feedback:

1. When a video is opened by clicking a link in the video description, there is currently no way to go back to the original video.  When such link is clicked, a new instance of YouTubePlayerFragment should be opened.
2. The above observation is also applicable to playlists -- there is no way to go back to the video when a playlist link is tapped.
3. We should expand this service such that it also handles channel links (i.e. the channel browser is opened if the user taps on a channel link).  [https://www.youtube.com/watch?v=_GzE99AmAQU has a link to a channel].

atomjack commented:
@ram-on I finally got around to fixing this branch up. It addresses all three points you made. I rebased it due to build.gradle conflicts, so make sure you erase your local copy.

ram-on commented:
Hello @atomjack, I'll tested it out in the upcoming days

ram-on commented:
Hi @atomjack, I've tested out and the only issue that needs rectification is the follows:
1. Go to https://youtu.be/tugFFhML7VY
2. Open the video description and tap the first link (https://www.youtube.com/watch?v=fAgUIxLcC6s&feature=youtu.be)
3. The following error should be displayed:  "Could not download channel information!"

Btw I'm currently working on a second version of YouTube player (YouTubePlayerV2Fragment), it would be nice to port that functionality to YouTubePlayerV2Fragment.

atomjack commented:
Hey @ram-on, thanks, I've added a fix for that url to be clicked on. It was a bug in the video matching regular expression. I've also rebased to resolve the build.gradle conflict so make sure you delete your local first.

atomjack commented:
@ram-on Oh yeah, re: YouTubePlayerV2Fragment, once you have a branch I can take a look at I'll let you know what needs to update in order for this functionality to support it.

atomjack commented:
@ram-on Cool. I actually did notice YouTubePlayerV2Fragment after I left that comment, and ran into some problems with the changes when trying to update my Chromecast branch. I've emailed you about it.

ram-on commented:
Thanks
